### ISSUE TITLE
Let's remove the ExceptionLog

### ISSUE THREAD

**Issue by jaimeiniesta** [jaimeiniesta/metainspector#154]
I'm thinking about removing completely the `ExceptionLog` on MetaInspector 5.0, and also the `warn_level` options, and instead let the exceptions be raised as they happen.

If we read the [README about exceptions](https://github.com/jaimeiniesta/metainspector#exception-handling) we see that we're advising not to use it, so let's be more opinionated and remove it completely.

I think this will lead to simpler, more maintainable code, removing [code like this](https://github.com/jaimeiniesta/metainspector/blob/master/lib/meta_inspector/request.rb#L41-L45) that is used just to get the `ExceptionLog` a chance to `:warn` or `:store` it, when the default action will be just to  [re-raise the exception](https://github.com/jaimeiniesta/metainspector/blob/master/lib/meta_inspector/exception_log.rb#L15).

Instead, I would document on the README what are the typical exceptions you're likely to encounter: timeouts, SSL, connection, encoding errors, etc., and let you handle it *outside* of MetaInspector.

I think that's way better than storing the errors and having to ask constantly for `page.ok?`, dealing with an inconsistent state of MetaInspector in the case of stored exceptions.

I will leave this open for discussion for some days before I go ahead and remove it, I would like to read your opinions!

What do you think?

**Comment by jaimeiniesta** [comment]
If we remove the ExceptionLog, I think we could:

a) Just have the exceptions raise as they happen. And document thoroughly what are the typical exceptions you can encounter: `Faraday::TimeoutError, Faraday::Error::ConnectionFailed, Faraday::SSLError, URI::InvalidURIError`... or

b) Simplify this and encapsulate those exceptions into our own generic MetaInspector exceptions, which would be basically 2, `MetaInspector::RequestError` for errors related with the fetching of the page, connection, SSL, timeouts... or `MetaInspector::ParserError` for errors raised during the parsing of the document. To still be able to give the exact details of the exception, we could embed the original exception using [nesty](https://github.com/skorks/nesty).

**Comment by jaimeiniesta** [comment]
Related conversation, this is when we introduced the ExceptionLog (although we already warned against its use and preferred to raise exceptions as they happen) https://github.com/jaimeiniesta/metainspector/issues/49

**Comment by jaimeiniesta** [comment]
What do you think @dentarg @jschwindt @mmmmmrob @gustavoguimaraes @hokaccha @gabceb @settinghead @oriolgual ?

**Comment by jaimeiniesta** [comment]
Another point is what good is having an `ExceptionLog` **inside** MetaInspector that only serves to catch and store any exception, if you could do that **outside** MetaInspector?

```ruby
@errors = []

begin
  page = MetaInspector.new('http://thispagedoesnotexistiswear.com')
rescue Exception => e
  puts "An exception happened, let's store it"
  @errors << e
end
```

This is simpler than having to check constantly for `page.ok`, and gives you more power than simply having the exceptions stored somewhere, for example you might want to treat in a different way the different kinds of exceptions, like retrying a number of times on connection errors with an exponential backoff, notifying externally only certain kinds of exceptions, etc.

**Comment by dentarg** [comment]
Just a quick response: I would prefer b), because it sounds easier to act on as user of MetaInspector.

I hope to get some time next week to think/write some more here.

**Comment by dentarg** [comment]
I'm wondering... is MetaInspector going to continue to list known exceptions that can happen, which would be great, because then I don't have to do that work, or are MetaInspector going to give up on that too? (Sounds like the later in https://github.com/jaimeiniesta/metainspector/pull/153#issuecomment-156751387)

**Comment by jaimeiniesta** [comment]
Yes, I prefer option b) as well, it will be easier to handle from the outside.

We're going to continue listing known exceptions.

**Comment by jschwindt** [comment]
In my case it would be enough to know if the whole process succeeded (request + parse) so any of both solutions would be ok. If I had to vote I'd choose b) because it's simpler for the user that doesn't have to know the inner details of the gem.

**Comment by jaimeiniesta** [comment]
Yes, this is what I'm working on, we can abstract all those exceptions into 2 or 3 generic errors and still have all the details, like this:

```ruby
2.2.3 :005 > MetaInspector.new('http://dasdadasd.com')
MetaInspector::RequestError: getaddrinfo: nodename nor servname provided, or not known
	from /Users/jaime/code/sitevalidator/metainspector/lib/meta_inspector/request.rb:40:in `rescue in response'
	from /Users/jaime/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:879:in `initialize': getaddrinfo: nodename nor servname provided, or not known (Faraday::ConnectionFailed)
	from /Users/jaime/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:879:in `open'
	from /Users/jaime/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:879:in `block in connect'
	from /Users/jaime/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/timeout.rb:73:in `timeout'
	from /Users/jaime/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:878:in `connect'
	from /Users/jaime/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:863:in `do_start'
	from /Users/jaime/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:852:in `start'
	from /Users/jaime/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:1375:in `request'
	from /Users/jaime/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:1133:in `get'
	from /Users/jaime/.rvm/gems/ruby-2.2.3/gems/faraday-0.9.2/lib/faraday/adapter/net_http.rb:80:in `perform_request'
	from /Users/jaime/.rvm/gems/ruby-2.2.3/gems/faraday-0.9.2/lib/faraday/adapter/net_http.rb:40:in `block in call'
	from /Users/jaime/.rvm/gems/ruby-2.2.3/gems/faraday-0.9.2/lib/faraday/adapter/net_http.rb:87:in `with_net_http_connection'
	from /Users/jaime/.rvm/gems/ruby-2.2.3/gems/faraday-0.9.2/lib/faraday/adapter/net_http.rb:32:in `call'
	from /Users/jaime/.rvm/gems/ruby-2.2.3/gems/faraday-cookie_jar-0.0.6/lib/faraday/cookie_jar.rb:22:in `call'
	from /Users/jaime/.rvm/gems/ruby-2.2.3/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response/follow_redirects.rb:76:in `perform_with_redirection'```

**Comment by jaimeiniesta** [comment]
I've started working on this: https://github.com/jaimeiniesta/metainspector/pull/155/files

**Comment by dentarg** [comment]
> We're going to continue listing known exceptions.

:+1: 

Great, I will continue report them/add PRs with them. :)

**Comment by dentarg** [comment]
I guess this will also solve problems like this, where `@exception_log` is nil for some reason (threading issues? we use [bunny](https://github.com/ruby-amqp/bunny) in our app)

```ruby
[2] pry(main)> MetaInspector.new("http://mjsantander.blogspot.com", warn_level: :store).ok?
=> true

[3] pry(main)> MetaInspector.new("http://mjsantander.blogspot.com", warn_level: :store).links.external
NoMethodError: undefined method `<<' for nil:NilClass
from /Users/dentarg/.gem/ruby/2.2.3/gems/metainspector-4.7.2/lib/meta_inspector/url.rb:88:in `rescue in normalized'

[4] pry(main)> wtf???
Exception: NoMethodError: undefined method `<<' for nil:NilClass
--
 0: /Users/dentarg/.gem/ruby/2.2.3/gems/metainspector-4.7.2/lib/meta_inspector/url.rb:88:in `rescue in normalized'
 1: /Users/dentarg/.gem/ruby/2.2.3/gems/metainspector-4.7.2/lib/meta_inspector/url.rb:86:in `normalized'
 2: /Users/dentarg/.gem/ruby/2.2.3/gems/metainspector-4.7.2/lib/meta_inspector/url.rb:53:in `url='
 3: /Users/dentarg/.gem/ruby/2.2.3/gems/metainspector-4.7.2/lib/meta_inspector/url.rb:15:in `initialize'
 4: /Users/dentarg/.gem/ruby/2.2.3/gems/metainspector-4.7.2/lib/meta_inspector/parsers/links.rb:37:in `new'
 5: /Users/dentarg/.gem/ruby/2.2.3/gems/metainspector-4.7.2/lib/meta_inspector/parsers/links.rb:37:in `block in external'
 6: /Users/dentarg/.gem/ruby/2.2.3/gems/metainspector-4.7.2/lib/meta_inspector/parsers/links.rb:37:in `select'
 7: /Users/dentarg/.gem/ruby/2.2.3/gems/metainspector-4.7.2/lib/meta_inspector/parsers/links.rb:37:in `external'
 8: (pry):3:in `__pry__'
```

Around metainspector-4.7.2/lib/meta_inspector/url.rb:88

```
    # Normalize url to deal with characters that should be encoded,
    # add trailing slash, convert to downcase...
    def normalized(url)
      Addressable::URI.parse(url).normalize.to_s
    rescue Addressable::URI::InvalidURIError => e
      @exception_log << e
      nil
    end
```

**Comment by jaimeiniesta** [comment]
@dentarg sure!

**Comment by jaimeiniesta** [comment]
I've removed the ExceptionLog, on a temporary 5-0-0 branch that will soon be merged into master:

https://github.com/jaimeiniesta/metainspector/tree/5-0-0

Please give it a try!

**Comment by jaimeiniesta** [comment]
I've released this as 5.0.0.rc1

**Comment by jaimeiniesta** [comment]
Just released 5.0.0, it's official, we don't have an ExceptionLog anymore :)

**Comment by robguilfoyle** [comment]
@jaimeiniesta, @dentarg, @jschwindt great job gentlemen on the release! I wish I had more time to contribute to this repo!
